### PR TITLE
Apply code style to project

### LINKED ISSUE #3123: Replace Flake8 & isort with Ruff
## `invoke lint` yields error with flake8 6.0.0

I just tried to run `invoke lint` on the code. This resulted in `flake8` not being installed first (not installed by `invoke install`, followed by:

> flake8: error: unrecognized arguments: --diff

This is related to flake8 6.0.0 dropping the `--diff` option, see https://flake8.pycqa.org/en/latest/release-notes/6.0.0.html#backwards-incompatible-changes. Using `pip install "flake8<6.0.0"` works, but the functionality should probably be fixed to not rely on old versions any more.

### PR DESCRIPTION
**Pull request by justinmayer**
- Remove upper version caps
- Update the minimum version of most of Pelican's runtime deps
- Replace Black with Ruff as a formatter for Pelican
- Add a cache step to the docs CI task so that the docs can be
  downloaded and inspected.
- Run pre-commit on all files during CI test job

Resolves #3217
Resolves #3227
Closes #3123

### CODE REVIEW

**Comment by justinmayer** [comment]
Assuming the SHA doesn't change for the commit titled _[Apply code style to project via: ruff format .](https://github.com/getpelican/pelican/pull/3231/commits/cabdb26cee66e1173cf16cb31d3fe5f9fa4392e7)_, before this PR is merged, the following commit hash should be added to the `.git-blame-ignore-revs` file so that blame will be correct: `cabdb26cee66e1173cf16cb31d3fe5f9fa4392e7`

**Review by offbyone — APPROVED** [review]
Ship it!

**Comment by boxydog** [comment]
Oof, the double quotes. I don't object, but this really rules out any automated resolution of any existing PRs. But, I approve of forward motion!

File: pyproject.toml [hunk 1]
```diff
@@ -76,28 +76,28 @@ test = "invoke tests"
...
+    "pygments>=2.16.1",
+    "pytest>=7.4.3",
+    "pytest-cov>=4.1.0",
+    "pytest-sugar>=0.9.7",
+    "pytest-xdist>=3.3.1",
+    "tox>=4.11.3",
+    "flake8>=6.1.0",
+    "flake8-import-order>=0.18.2",
+    "invoke>=2.2.0",
+    "isort>=5.12.0",
```
Comment (lioman) [inline pyproject.toml:97]: Shouldn't `isort` be removed then?
```suggestion
```

Comment (justinmayer) [reply pyproject.toml:97]: I believe so, and for that matter I think Flake8 and Black should also be removed, right @offbyone?

Comment (boxydog) [reply pyproject.toml:97]: I am not @offbyone, but I don't see black or isort in .pre-commit-config.yaml, so I agree.

Comment (offbyone) [reply pyproject.toml:97]: Correct; I left them in on the chance that someone ran them by hand, but yeah, they should be yeeted.

**Review by pauloxnet — APPROVED** [review]
LGTM

**Comment by justinmayer** [comment]
Many thanks to @boxydog, @offbyone, @lioman, and @pauloxnet for helping with this endeavor. Much appreciated! 🤗

**Pull request merged**